Dockerfile

```diff
@@ -18,59 +18,109 @@ RUN apk --no-cache add \
     && rm -rf /var/cache/apk/*
 
 # Setup repo
-COPY . ${GOPATH}/src/code.gitea.io/gitea
 WORKDIR ${GOPATH}/src/code.gitea.io/gitea
 
+COPY Makefile .
+
+# Download Golang Modules
+COPY go.mod .
+COPY go.sum .
+
+RUN make deps-backend
+
+# Download NPM Packages
+COPY package.json .
+COPY package-lock.json .
+
+RUN make deps-frontend
+
+COPY . .
```

Comment:
We could even further optimize by explicitly copying only what is needed here, which would enable us to remove `.dockerignore` as well. It's a bit risky though in case someone adds new root files/folders and forgets to add them here, but overall I'm in favor of it. Wonder what others think.

Reply:
I think you would still want to keep the .dockerignore as this stops the files being moved into the docker build context which is created before the first COPY command, which helps to reduce docker build times.

I use explicit COPYing elsewhere to help reduce build times in other projects. Especially nice to separate frontend and backend compilation, for an even higher chance of caching.

As for the risk of forgetting to add new files/folders to the Dockerfile, I don't think this would be too bad as the docker build workflows should flag this issue pretty quickly.